orttraining/orttraining/core/session/training_session.cc

```diff
@@ -125,7 +125,16 @@ Status TrainingSession::ConfigureForTraining(
                                          config.distributed_config.data_parallel_size,
                                          config.distributed_config.horizontal_parallel_size});
 
-  ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph());
+  // derive actual set of weights to train
+  std::unordered_set<std::string> weight_names_to_train =
+      !config.weight_names_to_train.empty()
+          ? config.weight_names_to_train
+          : GetTrainableModelInitializers(config.immutable_weights);
+  for (const auto& weight_name_to_not_train : config.weight_names_to_not_train) {
+    weight_names_to_train.erase(weight_name_to_not_train);
+  }
+
+  ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph(weight_names_to_train));
```

Comment:
if a Constant or Initializer is folded and removed from the graph, will that cause problem when saving state dict from PT frontend side?
@SherlockNoMad,

Reply:
I don't know for sure, but my guess is that it won't cause problem. 
Do we need to store non-trainable weights in the checkpoint files? Are they part of state dict ? 

---
In reply to: [409510819](https://github.com/microsoft/onnxruntime/pull/3544#discussion_r409510819) [](ancestors = 409510819)

Reply:
seems so, a PT parameter with requires_grad=False, appears in state_dict.

Reply:
For requires_grad=False cases are those parameters in frozen_weights, we can pass them by ort_parameters.weights_not_to_train and add them to the excluded set. Currently in ort_trainer.py we don't pass this information from frontend to ORT.

Reply:
As synced offline, let PT side handles this case.